gui/mechanisms.lua

```diff
@@ -0,0 +1,601 @@
+--Unlink and free mechanisms from buildings
+
+local gui = require("gui")
+local guidm = require("gui.dwarfmode")
+local widgets = require("gui.widgets")
+
+saved_mode = saved_mode or 0
+
+local function valid_building(b) --linkable
+    if not b then
+        return false
+    end
+
+    local t = b._type
+    if t == df.building_trapst then
+        if b.trap_type == df.trap_type.Lever or
+            b.trap_type == df.trap_type.PressurePlate or
+            b.trap_type == df.trap_type.TrackStop then
+                return true
+        end
+    elseif t == df.building_bars_floorst or
+        t == df.building_bars_verticalst or
+        t == df.building_bridgest or
+        t == df.building_cagest or
+        t == df.building_chainst or
+        t == df.building_doorst or
+        t == df.building_floodgatest or
+        t == df.building_gear_assemblyst or
+        t == df.building_grate_floorst or
+        t == df.building_grate_wallst or
+        t == df.building_hatchst or
+        t == df.building_supportst or
+        t == df.building_weaponst then
+            return true
+    end
+
+    return false
+end
+
+local function mech_iter(b) --iterate freeable mechanisms
+    local t = b.contained_items
+    local i_cur = 1 --index 0 is always building component
+    local i_max = #t - 1
+    return function()
+        for i = i_cur, i_max do
+            if t[i].item._type == df.item_trappartsst and t[i].item.flags.in_building then
+                i_cur = i + 1
+                return t[i].item
+            end
+        end
+    end
+end
+
+local function get_trigger_index(m) --return gref index or nil
+    for k, gref in ipairs(m.general_refs) do
+        if gref._type == df.general_ref_building_triggerst or
+            gref._type == df.general_ref_building_triggertargetst then
+                return k
+        end
+    end
+end
+
+local function get_mech_link(m) --return target building or nil
+    local x = get_trigger_index(m)
+    x = x and df.building.find(m.general_refs[x].building_id) or NULL
+    return not df.isnull(x) and x or nil
+end
+
+local function unlink_building(b, m) --clean up extra stuff
+    if b._type == df.building_trapst then
+        for k, v in ipairs(b.linked_mechanisms) do
+            if v == m then
+                b.linked_mechanisms:erase(k)
+                break
+            end
+        end
+    elseif b._type == df.building_doorst or b._type == df.building_hatchst then
+        b.door_flags.forbidden = false
+        b.door_flags.operated_by_mechanisms = false
+
+        local mb = dfhack.maps.getTileBlock(b.pos)
+        if mb then
+            mb.flags.repath_on_melt = true --fix pathing
```

Comment:
Is there a better way to clean up a tile's pathability than this?

Reply:
You'll likely need to set `df.global.world.reindex_pathfinding = true` to have DF recalculate the pathability flood fill

Reply:
Doesn't seem sufficient.